### PR TITLE
Init CUDA instead of faking memory stats

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121698
* #121707

This is very confusing when checking for memory usage and allocations are only happening using C API. We should change it to a warning/error or just init cuda. Codepaths that run on non-CUDA environments shouldn't call into these functions in the first place

cc @ptrblck @msaroufim

### CODE REVIEW

File: torch/cuda/memory.py [hunk 1]
```diff
@@ -158,8 +158,8 @@ def empty_cache() -> None:
         of GPU memory in certain cases. See :ref:`cuda-memory-management` for
         more details about GPU memory management.
     """
-    if is_initialized():
-        torch._C._cuda_emptyCache()
+    _lazy_init()
+    torch._C._cuda_emptyCache()
```
Comment (jansel) [inline torch/cuda/memory.py:162]: Why is this needed?  Won't the cache be empty if cuda is not initialized?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/cuda/memory.py (hunk 1 above)
Comment (xmfan) [reply torch/cuda/memory.py:162]: `is_initialized` only checks python side. 

So we could have a python script that calls some C++ pybind, where only C++ does cuda operations. Once C++ is done, we want to clear the cache from python side, but this is_initialized() check will still be false, and we will silently ignore the call

**Review by jansel — APPROVED** [review]

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but I think it breaks inductor CPU tests https://hud.pytorch.org/pytorch/pytorch/commit/5b90074540577267c29f5f784be123ee54f6491d' -c nosignal

The error is

```
  File "benchmarks/dynamo/timm_models.py", line 375, in <module>
    timm_main()
  File "benchmarks/dynamo/timm_models.py", line 371, in timm_main
    main(TimmRunner())
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 3428, in main
    process_entry(0, runner, original_dir, args)
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 3385, in process_entry
    return maybe_fresh_cache(
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 1942, in inner
    return fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 3940, in run
    runner.run_one_model(
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 2801, in run_one_model
    status = self.check_accuracy(
  File "/var/lib/jenkins/workspace/benchmarks/dynamo/common.py", line 2393, in check_accuracy
    torch.cuda.empty_cache()
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/cuda/memory.py", line 161, in empty_cache
    _lazy_init()
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/cuda/__init__.py", line 284, in _lazy_init
    raise AssertionError("Torch not compiled with CUDA enabled")
AssertionError: Torch not compiled with CUDA enabled
```

**Comment by Vargol** [comment]
Did this change make it in the current nightly ( 2.3.0.dev20240314) before it was revoked  ?

Accelerate is currently broken on anything non-CUDA when using the nightly, seems there stacks of code out there that only works on non CUDA environments due to CUDA stuff like empty_cache and memory_allocation stats being "no-ops" on non-CUDA platforms.

**Comment by xmfan** [comment]
@Vargol yes, looks this made it to 3/14 nightly. are you able to use yesterday's for today?

**Comment by Vargol** [comment]
Yes I can, or wait to test the next nightly.
Its not a big issue for me I just wanted the 3/14 nightly as it has got a MPS (technically metal) bf16 matmul fix in I wanted test in situ (my test code works so I'd expect it works in the app I'm using too) and also didn't want to raise a issue if the commit had been already reverted.

**Comment by xmfan** [comment]
This change is fairly self-contained, you might be able to test today if you're willing to undo these changes in your local pytorch installation's source. Here's an example using anaconda v.4.14.0: 
`<PATH TO .conda>/envs/<NAME OF CONDA ENV>/lib/python3.10/site-packages/torch/cuda/memory.py` e.g.
`~/.conda/envs/nightly/lib/python3.10/site-packages/torch/cuda/memory.py`

File: torch/cuda/memory.py (hunk 1 above)
Comment (ptrblck) [reply torch/cuda/memory.py:162]: @xmfan Could you give a valid use case where a C++ extension is using device memory and PyTorch is still able to clear it without initializing a CUDA context?

**Review by ptrblck — CHANGES_REQUESTED** [review]

File: torch/cuda/memory.py (hunk 1 above)
Comment (xmfan) [reply torch/cuda/memory.py:162]: I don't think there is such a case, (before this PR) PyTorch doesn't initialize a CUDA context and silently ignores the empty_cache, so the C++ extension will continue to use device memory

This PR was mainly to address empty_cache silent failure, but had to be reverted due to being called from CPU-only code in depending libraries
